rubix-tests/src/test/java/com/qubole/rubix/tests/TestGenerationNumber.java

```diff
@@ -0,0 +1,279 @@
+  private static final long TEST_LAST_MODIFIED = 1514764800; // 2018-01-01T00:00:00
+  private static final long TEST_FILE_LENGTH = 5000;
+  private static final String testDirectory = testDirectoryPrefix + "dir0";
+
+  private static final Log log = LogFactory.getLog(TestGenerationNumber.class);
+
+  @BeforeClass
+  public static void setupClass() throws IOException
+  {
+    log.info(testDirectory);
+    Files.createDirectories(Paths.get(testDirectory));
+  }
+
+  @AfterClass
+  public static void tearDownClass() throws IOException
+  {
+    Files.walkFileTree(Paths.get(testDirectory), new DeleteFileVisitor());
+    Files.deleteIfExists(Paths.get(testDirectory));
+  }
+
+  @BeforeMethod
+  public void setup() throws Exception
+  {
+    CacheConfig.setOnMaster(conf, true);
+    CacheConfig.setIsStrictMode(conf, true);
+    CacheConfig.setBookKeeperServerPort(conf, 3456);
+    CacheConfig.setDataTransferServerPort(conf, 2222);
+    CacheConfig.setBlockSize(conf, blockSize);
+    CacheConfig.setCacheDataDirPrefix(conf, testDirectoryPrefix + "dir");
+    CacheConfig.setMaxDisks(conf, 1);
+    CacheConfig.setIsParallelWarmupEnabled(conf, false);
+    CacheConfig.setCleanupFilesDuringStart(conf, false);
+    bookKeeperFactory = new BookKeeperFactory();
+    startServer();
+    // Populate File
+    DataGen.populateFile(backendFileName);
+    inputStream = createCachingStream(conf);
+    log.info("BackendPath: " + backendPath);
+  }
+
+  private CachingInputStream createCachingStream(Configuration conf)
+      throws IOException
+  {
+    FileSystem localFileSystem = new RawLocalFileSystem();
+    Path backendFilePath = new Path(backendFileName);
+    localFileSystem.initialize(backendFilePath.toUri(), new Configuration());
+    CacheConfig.setBlockSize(conf, blockSize);
+
+    // This should be after server comes up else client could not be created
+    return new CachingInputStream(backendPath, conf,
+        new CachingFileSystemStats(), ClusterType.TEST_CLUSTER_MANAGER,
+        new BookKeeperFactory(), localFileSystem,
+        CacheConfig.getBlockSize(conf), null);
+  }
+
+  @Test
+  void testReadWithInvalidations() throws Exception {
+    testReadWithInvalidationHelper();
+    CacheConfig.setIsParallelWarmupEnabled(conf, true);
+    restartServer();
+    testReadWithInvalidationHelper();
+  }
+
+  void testReadWithInvalidationHelper() throws TException, InterruptedException {
+    final int readOffset = 0, readLength = 2000;
+    RetryingPooledBookkeeperClient client = getBookKeeperClient();
+    Thread invalidateRequest = new Thread(new Runnable() {
+      @Override
+      public void run() {
+        for (int i = 1; i < 15; i++)
+        {
+          try {
+            client.invalidateFileMetadata(backendPath.toString());
+            Thread.sleep(1000);
+          } catch (Exception e) {
+            throw new RuntimeException(e);
+          }
+        }
+      }
+    });
+    Thread readrequest = new Thread(new Runnable() {
```

Comment:
Start multiple reader threads, each reading different range of file and in each cycle validate that right data is read